### ISSUE TITLE
FSelectorRcpp: Faster information.gain, gain.ratio and symmetrical uncertainty filters

### ISSUE THREAD

**Issue by studerus** [mlr-org/mlr#1677]
Some of the feature selection methods of the FSelector package have now been reimplented with Rcpp in the FSelectorRcpp package. The package has just been released on CRAN. As you can see in the following examples, the new implentation runs about 100x faster:

```
library(mlbench)
library(microbenchmark)
library(FSelector)
library(FSelectorRcpp)
data(HouseVotes84)
dat = na.omit(HouseVotes84)

microbenchmark(old = information.gain(Class~., dat),
               new = information_gain(Class~., dat), times = 10)

microbenchmark(old = gain.ratio(Class~., dat),
               new = information_gain(Class~., dat, type = 'gainratio'), times = 10)

microbenchmark(old = symmetrical.uncertainty(Class~., dat),
               new = information_gain(Class~., dat, type = 'symuncert'), times = 10)

```

I suggest that we replace these filters with the implentation in FSelectorRcpp.

**Comment by berndbischl** [comment]
> I suggest that we replace these filters with the implentation in FSelectorRcpp.

while i completely see your point, and also thank you for the benchmark, i would like to raise an important point (which we *already* screwed up, IMHO):

if we "naively" simply replace the calls into FSelector to FSelectorRCPP are we *sure* we dont break reproducibility behind peoples' back? in other words: does FSelectorRcpp really produce exactly the same numbers as FS?

**Comment by berndbischl** [comment]
> which we already screwed up, IMHO

i mean this here

```
> listFilterMethods()[,-3]                                                                                                                
                           id         package
1                  anova.test           Rfast
2                    carscore            care
3          cforest.importance           party
4                 chi.squared       FSelector
5                  gain.ratio       FSelector
6            information.gain       FSelector
7                kruskal.test                
8          linear.correlation           Rfast
9                        mrmr           mRMRe
10                       oneR       FSelector
11     permutation.importance                
12    randomForest.importance    randomForest
13      randomForestSRC.rfsrc randomForestSRC
14 randomForestSRC.var.select randomForestSRC
15           rank.correlation           Rfast
16                     relief       FSelector
17              rf.importance randomForestSRC
18               rf.min.depth randomForestSRC
19    symmetrical.uncertainty       FSelector
20                 univariate                
21     univariate.model.score                
22                   variance      
```

**Comment by berndbischl** [comment]
we now use Rfast for stuff, which was computed before with FSelector. 
and rf.importance is now computed with rfsrc, if used to be with the normal randomForest.

I would guess that the implemenations do not compute exactly the same numbers. for rf.importance the name is now even misleading for the user.

@jakob-r

**Comment by berndbischl** [comment]
@mllg suggested that the filter constructor should also take "package" as an argument to select.
and we only add new methods and possibly deprecate stuff we dont want to maintain anymore.

but we dont "overwrite"

**Comment by mllg** [comment]
Note that they seem to compare their calculated scores to FSelector.
+1 for integrating the Rcpp version, but with the mentioned argument.

**Comment by studerus** [comment]
The FSelector and FSelectorRcpp versions should give exactly the same results and it was true for my example. Hence, I don't see why cannot simply replace the older version. If we have a "package" argument, then the FSelectorRcpp package should at least be the default because I cannot imagine that I would ever want to use the slower version.

**Comment by berndbischl** [comment]
> should 

who is responsible for checking this? i mean, michel and i looked at their code and they seem to have 
unit tests, that ensure that at least on some examples.

but in general, i mean, assume this situation. we now switch to the Rcpp version. some filters actually dont produce the same output. this might be a bug, or a choice they made in FSrcpp. you ran your experiments with mlr 2.10. now for 2.12 we have silently behind your back switched to a new filter version. your experiments now return different results and are not reproducible anymore. 

maybe, i guess one could argue that we have documented in NEWS and you need to install 2.10 for full reproducibility? 

i am a bit on the fence here. I really would like to get away with the old java FS and i dont want to support 2 versions of the same thing. i just think we should worry about things like this. 

any other opinions? 
@larskotthoff

**Comment by berndbischl** [comment]
@ja-thomas @giuseppec

**Comment by larskotthoff** [comment]
I would say it's the responsibility of FSelectorRcpp to show that they produce the same results, or justify where they don't. I should probably contact the authors and ask them about it.

**Comment by berndbischl** [comment]
> would say it's the responsibility of FSelectorRcpp to show that they produce the same results, or justify where they don't

but that sidesteps the question about reproducibility a bit though. 

well i guess lets do the switch

**Comment by ja-thomas** [comment]
> maybe, i guess one could argue that we have documented in NEWS and you need to install 2.10 for full reproducibility?

+1

Clearly state in NEWS that we switched out the package to calculate these measures. They should be the same but to guarantee reproducibility use mlr 2.10.

People who care about reproducibility should use docker or packrat anyways.

**Comment by berndbischl** [comment]
while we are at it, can we please discuss these names. they look very weird.

> 12    randomForest.importance    randomForest
> 13      randomForestSRC.rfsrc randomForestSRC
> 14 randomForestSRC.var.select randomForestSRC
> 17              rf.importance randomForestSRC
> 18               rf.min.depth randomForestSRC


12: a bit long, but clean
13: a bit weird that we use rfsrc, later we call  it importance. but could be argued it is package.function and ok
14: package.function. so ok?
17 - 18: "rf" is really misleading. improve?

**Comment by larskotthoff** [comment]
See mi2-warsaw/FSelectorRcpp#51.

**Comment by schiffner** [comment]
> 17 - 18: "rf" is really misleading. improve?

These two filters are deprecated, so we don't need to worry about them.

12-14 all fit the package.function naming scheme.
But since we do not generally use this scheme for all filters one could argue that sth. like rfsrc -> imporance and var.select -> minimal.depth would be more self-explanatory.

**Comment by engAnt** [comment]
good to know that someone else has ran into this (FSrcpp 100x faster than FS) to remove any doubt of my setup or limited mlr skills
shout out to the mlr authors and maintainers for thinking of ways to improve things as well

has anyone looked into contributing to the FSrcpp package to maybe add functionality so that it is mlr standards complaint in a way? just something i thought about after today while trying out FS and FSrcpp. I just started developing R packages this is why I have not started contributing to other packages just yet

**Comment by RandomGuessR** [comment]
Hi, just wondering whether anyone has had the chance to test FSelectorRcpp against FSelector for reproducibility of results.

I tried it against a fairly wide dataset (the Santander value prediction challenge on Kaggle - [https://www.kaggle.com/c/santander-value-prediction-challenge] with 4,991 columns and 4,459 rows), and I struggled to get reproducible results for both information gain and symmetrical uncertainty.  For FSelectorRcpp, I tried both default options (i.e. equal = FALSE) and with equal = TRUE.  The outcomes I get is that some of the features get very similar attribute importance, while others get vastly different.

E.g. see this extract - the first and third features gets identical importance under both methods, others are different, and the last one here is 0 with the java-based version and the highest importance out of these using the rcpp version:

Feature  | FSelector | FSelectorRcpp_equal
-- | -- | --
X48df886f9 | 0.004739088 | 0.004739088
X0deb4b6a8 | 0 | 0.001443917
X34b15f335 | 0 | 0
a8cb14b00 | 0 | 0.000722185
X2f0771a37 | 0 | 0.001804959
X30347e683 | 0 | 0.010275108

The code is simple:
`IG_java <- information.gain(resp~.,DF)
IG_rcpp3 <- information_gain(resp ~., DF, equal = TRUE, threads = mythreads)
`
where resp is my target (response) and DF is a data frame with the predictor columns followed by the response column, built using `
 DF <- as.data.frame(cbind(pred, resp))
`

As mentioned, I also tried the default setting; that did not produce better results:
`IG_rcpp2 <- information_gain(resp ~., DF, threads = mythreads)`

I tried the matrix input format for the rcpp implementation; that produced the same outcome as the formula input (and different from the java implementation):
`IG_rcpp4 <- information_gain(x = pred, y = resp, equal = TRUE, threads = mythreads)`

All of these tests had the same pattern when testing with symmetrical uncertnainty.

When I use the two respective implementations (java & rcpp) as a feature selector for a model, the outcomes I get are vastly different, with the java-based method hugely outperforming the rcpp method on this dataset (with all other modelling parameters kept constant, of course).

Other than that indeed the rcpp version runs ~100x faster on my machine - ~224 seconds (4 minutes) for the java-based version and ~2 seconds for the rcpp version on 8 threads.

I searched for any performance comparisons between the two implementations, but did not find any other than this discussion thread.

Thanks for looking into this!

**Comment by zzawadz** [comment]
@RandomGuessR 

`FSelectorRcpp` treats integer columns in a different way than the `FSelector`, so this might be the source of the difference in the results.

In FSelectorRcpp integers are treated in the same way as factors, so we do not use discretization before calculating the information gain. `FSelector` handles integers as ordinal numeric vectors, so it applies the discretization.

It is shown in the example below, the result for `FSelector` is the same for types integer and numeric. For `FSelectorRcpp` the results are different.

See the code below:

``` r
library(FSelectorRcpp)
library(FSelector)

dt <- iris[,c(1,5)]
dt[["SepLenInteger"]] <- as.integer(dt$Sepal.Length)
dt[["SepLenNumeric"]] <- as.numeric(as.integer(dt$Sepal.Length))
dt <- dt[,-1] # remove Sepal.Length column

FSelector::information.gain(Species ~ ., dt)
#>               attr_importance
#> SepLenInteger       0.3216773
#> SepLenNumeric       0.3216773
FSelectorRcpp:::information_gain(Species ~ ., dt)
#>      attributes importance
#> 1 SepLenInteger  0.3922533
#> 2 SepLenNumeric  0.3216773
```

I prefer the `FSelectorRcpp` approach because some people might use integers to encode factor levels (e.g., the day of a week, genders, and so on), so it's better not to use discretization here. However, I'll add a parameter to control this behavior (something like `integer2numeric`), to make the user aware of this problem.

**Comment by zzawadz** [comment]
I also think that the issues with `FSelectorRcpp` are not the problems of `mlr`, so they should be posted on the `FSelectorRcpp` page - https://github.com/mi2-warsaw/FSelectorRcpp/issues

**Comment by RandomGuessR** [comment]
@zzawadz  - agree, apologies, when I did a search I initially only found this thread so followed up here, only later found the other thread and posted there as well.  Sorry, new to the place.
Thanks again for the help!

**Comment by pat-s** [comment]
Implemented in #2140.
